### PR TITLE
Add map result support in neural search for non text embedding models

### PR DESCRIPTION
**Pull request by zane-neo**
### Description
Neural search only support text embedding model result which are usually in List<List<Float>> structure, for other models that returns non vector list like OpenAI, SPLADE etc, there's no available functions to get extract the json object result. This PR is to add several new functions to support these cases, the response are in Map<String, ?> structure and upper layer can extract more information from the map based on their purpose.

### Issues Resolved
https://github.com/opensearch-project/neural-search/issues/260

### Check List
- [ ] New functionality includes testing.
    - [ ] All tests pass
- [ ] New functionality has been documented.
    - [ ] New functionality has javadoc added
- [ ] Commits are signed as per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by navneet1v** [comment]
@zane-neo lets not skip the change log.

Also, motivation is not clear for this change. Can you add a github issue and talk about why we need this change? Is there any new feature that we are building which is going to use the new responses?

File: src/main/java/org/opensearch/neuralsearch/ml/MLCommonsClientAccessor.java [hunk 1]
```diff
@@ -144,4 +174,19 @@ private List<List<Float>> buildVectorFromResponse(MLOutput mlOutput) {
...
+    private Map<String, ?> buildMapResultFromResponse(MLOutput mlOutput) {
+        final ModelTensorOutput modelTensorOutput = (ModelTensorOutput) mlOutput;
+        final List<ModelTensors> tensorOutputList = modelTensorOutput.getMlModelOutputs();
+        if (CollectionUtils.isEmpty(tensorOutputList)) {
+            log.error("No tensor output found!");
+            return null;
+        }
+        List<ModelTensor> tensorList = tensorOutputList.get(0).getMlModelTensors();
+        if (CollectionUtils.isEmpty(tensorList)) {
+            log.error("No tensor found!");
```
Comment (navneet1v) [inline src/main/java/org/opensearch/neuralsearch/ml/MLCommonsClientAccessor.java:186]: lets make this error message more understandable and with actions what happened wrong which resulted in this error

File: src/main/java/org/opensearch/neuralsearch/ml/MLCommonsClientAccessor.java [hunk 2]
```diff
@@ -144,4 +174,19 @@ private List<List<Float>> buildVectorFromResponse(MLOutput mlOutput) {
         return vector;
     }
 
+    private Map<String, ?> buildMapResultFromResponse(MLOutput mlOutput) {
+        final ModelTensorOutput modelTensorOutput = (ModelTensorOutput) mlOutput;
+        final List<ModelTensors> tensorOutputList = modelTensorOutput.getMlModelOutputs();
+        if (CollectionUtils.isEmpty(tensorOutputList)) {
+            log.error("No tensor output found!");
```
Comment (navneet1v) [inline src/main/java/org/opensearch/neuralsearch/ml/MLCommonsClientAccessor.java:181]: lets make this error message more understandable and with actions what happened wrong which resulted in this error

File: src/main/java/org/opensearch/neuralsearch/ml/MLCommonsClientAccessor.java [hunk 3]
```diff
@@ -144,4 +174,19 @@ private List<List<Float>> buildVectorFromResponse(MLOutput mlOutput) {
         return vector;
     }
 
+    private Map<String, ?> buildMapResultFromResponse(MLOutput mlOutput) {
+        final ModelTensorOutput modelTensorOutput = (ModelTensorOutput) mlOutput;
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/neuralsearch/ml/MLCommonsClientAccessor.java:178]: should we check the type before casting?

**Comment by zane-neo** [comment]
> @zane-neo lets not skip the change log.
> 
> Also, motivation is not clear for this change. Can you add a github issue and talk about why we need this change? Is there any new feature that we are building which is going to use the new responses?

Sure, created this issue: https://github.com/opensearch-project/neural-search/issues/260. Yes, the SPLADE feature needs to use the new response.

**Comment by navneet1v** [comment]
@zane-neo as this is a new feature, and will probably go through multiple iterations before it is going to be released. Hence, for all new features let's not merge the changes directly in main. Lets merge keep reviewing the changes and merge them in a feature branch. Let me create a feature branch for this new feature.

So the process will go like this:
1. Neural Search plugin maintainer will cut a feature branch from main branch.
2. Contributors working on Sparse Vectors will raise the PR against the feature branch. 
3. Once the PR is reviewed it will go in the feature branch.
4. Once all the changes are done and performance testing is done, commits of feature branch will be merged in the main branch.

Please let me know if there is any further questions.

cc: @vamshin

**Comment by navneet1v** [comment]
@zane-neo Feature branch for sparse vector support: https://github.com/opensearch-project/neural-search/tree/feature/sparseVectorSupport 

Please use this branch for this and all future PRs related to SparseVector Support.

**Comment by zane-neo** [comment]
Closing this PR since raised another PR to feature branch: https://github.com/opensearch-project/neural-search/pull/270

**Pull request closed without merging**